File: 0312-burst-balloons/0312-burst-balloons.cpp
class Solution {
public:
    int dp[505][505];
    int solve(vector<int>& nums, int l, int r)
    {
        if(l > r) return 0;
        if(dp[l][r] != -1) return dp[l][r];
        
        int ans = INT_MIN;
        for(int i = l; i <= r; i++)
        {
            int x = nums[l-1] * nums[i] * nums[r+1];
            
            int lv = solve(nums, l, i-1);
            int rv = solve(nums, i+1, r);
            
            ans = max(ans, x + lv + rv);
        }
        return dp[l][r] = ans;
    }
    int maxCoins(vector<int>& nums) 
    {
        nums.insert(nums.begin(), 1);
        nums.push_back(1);
        
        memset(dp, -1, sizeof(dp));
        
        return solve(nums, 1, nums.size()-2);
    }
};

/*

[3,1,5,8]



*/

File: 0312-burst-balloons/0312-burst-balloons.java
class Solution {
    int F(int[] nums, int i, int j, int[][] dp) {
        if(i > j) return 0;
        if(dp[i][j] != -1) return dp[i][j];

        int mx = Integer.MIN_VALUE;
        for(int k = i; k <= j; k++)
        {
            int ans = nums[i-1] * nums[k] * nums[j+1];
            int value = F(nums, i, k-1, dp) + F(nums, k+1, j, dp);

            mx = Math.max(mx, ans + value);
        }

        return dp[i][j] = mx;
    }
    public int maxCoins(int[] nums) {
        int[] arr = new int[nums.length+2];
        int n = arr.length;
        
        arr[0] = arr[n-1] = 1;

        int[][] dp = new int[n+1][n+1];
        for(int i = 0; i <= n; i++) {
            for(int j = 0; j <= n; j++) {
                dp[i][j] = -1;
            }
        }

        for(int i = 1; i < n-1; i++) {
            arr[i] = nums[i-1];
        }
        
        return F(arr, 1, n-2, dp);
    }
}